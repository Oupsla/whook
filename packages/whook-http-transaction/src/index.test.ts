import StreamTest from 'streamtest';
import YError from 'yerror';
import initHTTPTransaction, { WhookHandler, WhookResponse } from './index';
import { IncomingMessage } from 'http';

function streamifyBody(response) {
  return Object.assign({}, response, {
    body: StreamTest.v2.fromChunks([JSON.stringify(response.body)]),
  });
}

describe('initHTTPTransaction', () => {
  const log = jest.fn();
  const time = jest.fn();
  const uniqueId = jest.fn();
  const delay = {
    create: jest.fn(),
    clear: jest.fn(),
  };

  beforeEach(() => {
    log.mockReset();
    time.mockReset();
    uniqueId.mockReset();
    delay.create.mockReset();
    delay.clear.mockReset();
  });

  test('should work', async () => {
    const httpTransaction = await initHTTPTransaction({
      log,
      time,
      delay,
      uniqueId,
    });

    expect(httpTransaction).toBeTruthy();
    expect({
      logCalls: log.mock.calls,
      timeCalls: time.mock.calls,
      uniqueIdCalls: uniqueId.mock.calls,
      delayCreateCalls: delay.create.mock.calls,
      delayClearCalls: delay.clear.mock.calls,
    }).toMatchSnapshot();
  });

  describe('httpTransaction', () => {
    const buildResponse = jest.fn();
    let res;
    let resBodyPromise;

    beforeEach(done => {
      buildResponse.mockReset();
      resBodyPromise = new Promise((resolve, reject) => {
        res = StreamTest.v2.toText((err, text) => {
          if (err) {
            reject(err);
            return;
          }
          resolve(text);
        });
        res.writeHead = jest.fn();
        done();
      });
    });

    test('should work', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockReturnValueOnce(new Promise(() => {}));
      delay.clear.mockResolvedValue(undefined);

      const httpTransaction = await initHTTPTransaction({
        log,
        time,
        delay,
        uniqueId,
      });
      const req = {
        connection: { encrypted: true },
        ts: 1000000,
        ip: '127.0.0.1',
        method: 'GET',
        url: '/v1/users/1?extended=true',
        headers: {
          'x-forwarded-for': '127.0.0.1',
        },
        socket: {
          bytesRead: 16,
          bytesWritten: 64,
        },
      };

      buildResponse.mockResolvedValueOnce({
        body: StreamTest.v2.fromChunks([
          JSON.stringify({
            id: 1,
            name: 'John Doe',
          }),
        ]),
        headers: {
          'Content-Type': 'application/json',
        },
        status: 200,
      });

      const { request, transaction } = await httpTransaction(
        (req as unknown) as IncomingMessage,
        res,
      );

      const response = await transaction.start(
        (buildResponse as unknown) as WhookHandler<any, any>,
      );

      await transaction.end(response);

      expect(request.url).toEqual('/v1/users/1?extended=true');
      expect(request.method).toEqual('get');
      expect(request.headers).toEqual({
        'x-forwarded-for': '127.0.0.1',
      });
      expect(request.body).toEqual(req);
      expect(res.writeHead.mock.calls).toEqual([
        [
          200,
          'OK',
          {
            'Content-Type': 'application/json',
            'Transaction-Id': '[id]',
          },
        ],
      ]);

      const text = await resBodyPromise;

      expect(text).toEqual('{"id":1,"name":"John Doe"}');
      expect({
        request,
        buildResponseCalls: buildResponse.mock.calls,
        logCalls: log.mock.calls,
        timeCalls: time.mock.calls,
        uniqueIdCalls: uniqueId.mock.calls,
        delayCreateCalls: delay.create.mock.calls,
        delayClearCalls: delay.clear.mock.calls,
      }).toMatchSnapshot();
    });

    test('should fail on timeout', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockResolvedValueOnce(undefined);
      delay.clear.mockResolvedValueOnce(undefined);

      let request;

      try {
        const httpTransaction = await initHTTPTransaction({
          log,
          time,
          delay,
          uniqueId,
        });
        const req = {
          connection: { encrypted: true },
          ts: 1000000,
          ip: '127.0.0.1',
          method: 'GET',
          url: '/v1/users/1?extended=true',
          headers: {
            'x-forwarded-for': '127.0.0.1',
          },
          socket: {
            bytesRead: 16,
            bytesWritten: 64,
          },
        };

        buildResponse.mockReturnValueOnce(new Promise(() => {}));

        const { request: _request, transaction } = await httpTransaction(
          (req as unknown) as IncomingMessage,
          res,
        );

        request = _request;

        await transaction.start(
          (buildResponse as unknown) as WhookHandler<any, any>,
        );

        await transaction.end({} as WhookResponse);

        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect(err.code).toEqual('E_TRANSACTION_TIMEOUT');
        expect(err.httpCode).toEqual(504);
        expect({
          request,
          logCalls: log.mock.calls,
          timeCalls: time.mock.calls,
          uniqueIdCalls: uniqueId.mock.calls,
          delayCreateCalls: delay.create.mock.calls,
          delayClearCalls: delay.clear.mock.calls,
        }).toMatchSnapshot();
      }
    });

    test('should fail with non-unique transaction id', async () => {
      time.mockReturnValue(new Date('2012-01-15T00:00:00.000Z').getTime());
      uniqueId.mockReturnValue('[id]');
      delay.create.mockReturnValueOnce(new Promise(() => {}));
      delay.clear.mockResolvedValue(undefined);

      try {
        const httpTransaction = await initHTTPTransaction({
          log,
          time,
          delay,
          uniqueId,
          TRANSACTIONS: { lol: {} },
        });
        const req = {
          connection: { encrypted: true },
          ts: 1000000,
          ip: '127.0.0.1',
          method: 'GET',
          url: '/v1/users/1?extended=true',
          headers: {
            'x-forwarded-for': '127.0.0.1',
            'transaction-id': 'lol',
          },
          socket: {
            bytesRead: 16,
            bytesWritten: 64,
          },
        };

        buildResponse.mockReturnValueOnce(Promise.resolve());

        const { transaction } = await httpTransaction(
          (req as unknown) as IncomingMessage,
          res,
        );

        await transaction
          .start((buildResponse as unknown) as WhookHandler<any, any>)
          .catch(transaction.catch)
          .then(streamifyBody)
          .then(transaction.end);

        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect(err.code).toEqual('E_TRANSACTION_ID_NOT_UNIQUE');
        expect(err.httpCode).toEqual(400);
        expect({
          timeCalls: time.mock.calls,
          uniqueIdCalls: uniqueId.mock.calls,
          delayCreateCalls: delay.create.mock.calls,
          delayClearCalls: delay.clear.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
