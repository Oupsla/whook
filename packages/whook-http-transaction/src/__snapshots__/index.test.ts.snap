// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initHTTPTransaction httpTransaction should fail on timeout 1`] = `
Object {
  "delayClearCalls": Array [],
  "delayCreateCalls": Array [
    Array [
      30000,
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "request": Object {
    "body": Object {
      "connection": Object {
        "encrypted": true,
      },
      "headers": Object {
        "x-forwarded-for": "127.0.0.1",
      },
      "ip": "127.0.0.1",
      "method": "GET",
      "socket": Object {
        "bytesRead": 16,
        "bytesWritten": 64,
      },
      "ts": 1000000,
      "url": "/v1/users/1?extended=true",
    },
    "headers": Object {
      "x-forwarded-for": "127.0.0.1",
    },
    "method": "get",
    "url": "/v1/users/1?extended=true",
  },
  "timeCalls": Array [
    Array [],
  ],
  "uniqueIdCalls": Array [
    Array [],
  ],
}
`;

exports[`initHTTPTransaction httpTransaction should fail with non-unique transaction id 1`] = `
Object {
  "delayClearCalls": Array [],
  "delayCreateCalls": Array [
    Array [
      30000,
    ],
  ],
  "timeCalls": Array [
    Array [],
  ],
  "uniqueIdCalls": Array [
    Array [],
  ],
}
`;

exports[`initHTTPTransaction httpTransaction should work 1`] = `
Object {
  "buildResponseCalls": Array [
    Array [],
  ],
  "delayClearCalls": Array [
    Array [
      Promise {},
    ],
  ],
  "delayCreateCalls": Array [
    Array [
      30000,
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
    Array [
      "info",
      Object {
        "endInBytes": 16,
        "endOutBytes": 64,
        "endTime": 1326585600000,
        "errored": false,
        "id": "[id]",
        "ip": "127.0.0.1",
        "method": "GET",
        "protocol": "http",
        "reqHeaders": Object {
          "x-forwarded-for": "127.0.0.1",
        },
        "resHeaders": Object {
          "Content-Type": "application/json",
        },
        "startInBytes": 16,
        "startOutBytes": 64,
        "startTime": 1326585600000,
        "statusCode": 200,
        "url": "/v1/users/1?extended=true",
      },
    ],
  ],
  "request": Object {
    "body": Object {
      "connection": Object {
        "encrypted": true,
      },
      "headers": Object {
        "x-forwarded-for": "127.0.0.1",
      },
      "ip": "127.0.0.1",
      "method": "GET",
      "socket": Object {
        "bytesRead": 16,
        "bytesWritten": 64,
      },
      "ts": 1000000,
      "url": "/v1/users/1?extended=true",
    },
    "headers": Object {
      "x-forwarded-for": "127.0.0.1",
    },
    "method": "get",
    "url": "/v1/users/1?extended=true",
  },
  "timeCalls": Array [
    Array [],
    Array [],
  ],
  "uniqueIdCalls": Array [
    Array [],
  ],
}
`;

exports[`initHTTPTransaction should work 1`] = `
Object {
  "delayClearCalls": Array [],
  "delayCreateCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ’± - HTTP Transaction initialized.",
    ],
  ],
  "timeCalls": Array [],
  "uniqueIdCalls": Array [],
}
`;
